File: crystal-ball/src/main/java/org/strangeforest/tcb/model/MatchOutcome.java
package org.strangeforest.tcb.model;

public class MatchOutcome extends DiffOutcome {

	MatchOutcome(double pServe, double pReturn, int bestOf) {
		this(pServe, pReturn, bestOf, finalSetTieBreaker(bestOf));
	}

	MatchOutcome(double pServe, double pReturn, int bestOf, boolean finalSetTieBreaker) {
		super(maxSets(bestOf), 1, setNo -> new SetOutcome(pServe, pReturn, finalSetTieBreaker || setNo < maxSets(bestOf)).pWin());
	}

	private static int maxSets(int bestOf) {
		switch (bestOf) {
			case 3: return 2;
			case 5: return 3;
			default: throw new IllegalStateException();
		}
	}

	private static boolean finalSetTieBreaker(int bestOf) {
		switch (bestOf) {
			case 3: return true;
			case 5: return false;
			default: throw new IllegalStateException();
		}
	}
}

File: crystal-ball/src/main/kotlin/org/strangeforest/tcb/model/MatchOutcome.kt
package org.strangeforest.tcb.model

class MatchOutcome @JvmOverloads constructor(
	pServe: Double,
	pReturn: Double,
	bestOf: Int,
	finalSetTieBreaker: Int? = finalSetTieBreaker(bestOf)
)
	: DiffOutcome(maxSets(bestOf), 1, { setNo -> SetOutcome(pServe, pReturn, if (setNo < maxSets(bestOf)) finalSetTieBreaker else 5).pWin() }
)
